File: api/src/main/java/org/openmrs/module/openhmis/cashier/api/model/Bill.java
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 2.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
 * the License for the specific language governing rights and
 * limitations under the License.
 *
 * Copyright (C) OpenHMIS.  All Rights Reserved.
 */
package org.openmrs.module.openhmis.cashier.api.model;

import java.math.BigDecimal;
import java.security.AccessControlException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.openmrs.BaseOpenmrsData;
import org.openmrs.Patient;
import org.openmrs.Provider;
import org.openmrs.api.context.Context;
import org.openmrs.module.openhmis.cashier.api.util.PrivilegeConstants;
import org.openmrs.module.openhmis.inventory.api.model.Item;
import org.openmrs.module.openhmis.inventory.api.model.ItemPrice;

/**
 * Model class that represents a list of {@link BillLineItem}s and {@link Payment}s created by a cashier for a patient.
 */
public class Bill extends BaseOpenmrsData {
	public static final long serialVersionUID = 0L;

	private Integer billId;
	private String receiptNumber;
	private Provider cashier;
	private Patient patient;
	private CashPoint cashPoint;
	private Bill billAdjusted;
	private BillStatus status;
	private List<BillLineItem> lineItems;
	private Set<Payment> payments;
	private Set<Bill> adjustedBy;
	private Boolean receiptPrinted = false;
	private String adjustmentReason;

	public String getAdjustmentReason() {
		return adjustmentReason;
	}

	public void setAdjustmentReason(String adjustmentReason) {
		this.adjustmentReason = adjustmentReason;
	}

	public Boolean isReceiptPrinted() {
		return receiptPrinted;
	}

	public void setReceiptPrinted(Boolean receiptPrinted) {
		this.receiptPrinted = receiptPrinted;
	}

	public BigDecimal getTotal() {
		BigDecimal total = BigDecimal.ZERO;

		if (lineItems != null) {
			for (BillLineItem line : lineItems) {
				if (line != null && !line.getVoided()) {
					total = total.add(line.getTotal());
				}
			}
		}

		return total;
	}

	public BigDecimal getTotalPayments() {
		BigDecimal total = BigDecimal.ZERO;

		if (payments != null) {
			for (Payment payment : payments) {
				if (payment != null && !payment.getVoided()) {
					total = total.add(payment.getAmount());
				}
			}
		}

		return total;
	}

	public BigDecimal getAmountPaid() {
		BigDecimal total = getTotal();
		BigDecimal totalPayments = getTotalPayments();

		return total.min(totalPayments);
	}

	@Override
	public Integer getId() {
		return billId;
	}

	@Override
	public void setId(Integer id) {
		billId = id;
	}

	public String getReceiptNumber() {
		return receiptNumber;
	}

	public void setReceiptNumber(String number) {
		this.receiptNumber = number;
	}

	public Provider getCashier() {
		return cashier;
	}

	public void setCashier(Provider cashier) {
		this.cashier = cashier;
	}

	public Patient getPatient() {
		return patient;
	}

	public void setPatient(Patient patient) {
		this.patient = patient;
	}

	public CashPoint getCashPoint() {
		return cashPoint;
	}

	public void setCashPoint(CashPoint cashPoint) {
		this.cashPoint = cashPoint;
	}

	public Bill getBillAdjusted() {
		return billAdjusted;
	}

	public void setBillAdjusted(Bill billAdjusted) {
		this.billAdjusted = billAdjusted;

		if (billAdjusted != null) {
			billAdjusted.setStatus(BillStatus.ADJUSTED);
		}
	}

	public BillStatus getStatus() {
		return status;
	}

	public void setStatus(BillStatus status) {
		this.status = status;
	}

	public List<BillLineItem> getLineItems() {
		return lineItems;
	}

	public void setLineItems(List<BillLineItem> lineItems) {
		this.lineItems = lineItems;
	}

	public BillLineItem addLineItem(Item item, ItemPrice price, int quantity) {
		if (item == null) {
			throw new NullPointerException("The item to add must be defined.");
		}
		if (price == null) {
			throw new NullPointerException("The item price must be defined.");
		}

		return addLineItem(item, price.getPrice(), price.getName(), quantity);
	}

	public BillLineItem addLineItem(Item item, BigDecimal price, String priceName, int quantity) {
		if (item == null) {
			throw new IllegalArgumentException("The item to add must be defined.");
		}
		if (price == null) {
			throw new IllegalArgumentException("The item price must be defined.");
		}

		BillLineItem lineItem = new BillLineItem();
		lineItem.setBill(this);
		lineItem.setItem(item);
		lineItem.setPrice(price);
		lineItem.setPriceName(priceName);
		lineItem.setQuantity(quantity);

		addLineItem(lineItem);

		return lineItem;
	}

	public void addLineItem(BillLineItem item) {
		if (item == null) {
			throw new NullPointerException("The list item to add must be defined.");
		}

		if (this.lineItems == null) {
			this.lineItems = new ArrayList<BillLineItem>();
		}

		this.lineItems.add(item);
		item.setBill(this);
	}

	public void removeLineItem(BillLineItem item) {
		if (item != null) {
			if (this.lineItems != null) {
				this.lineItems.remove(item);
			}
		}
	}

	public Set<Payment> getPayments() {
		return payments;
	}

	public void setPayments(Set<Payment> payments) {
		this.payments = payments;
	}

	public Payment addPayment(PaymentMode mode, Set<PaymentAttribute> attributes, BigDecimal amount,
	        BigDecimal amountTendered) {
		if (mode == null) {
			throw new NullPointerException("The payment mode must be defined.");
		}
		if (amount == null) {
			throw new NullPointerException(("The payment amount must be defined."));
		}

		Payment payment = new Payment();
		payment.setInstanceType(mode);
		payment.setAmount(amount);
		payment.setAmountTendered(amountTendered);

		if (attributes != null && attributes.size() > 0) {
			payment.setAttributes(attributes);

			for (PaymentAttribute attribute : attributes) {
				attribute.setOwner(payment);
			}
		}

		addPayment(payment);

		return payment;
	}

	public void addPayment(Payment payment) {
		if (payment == null) {
			throw new NullPointerException("The payment to add must be defined.");
		}

		if (this.payments == null) {
			this.payments = new HashSet<Payment>();
		}

		this.payments.add(payment);
		payment.setBill(this);

		this.checkPaidAndUpdateStatus();
	}

	public boolean checkPaidAndUpdateStatus() {
		if (this.getPayments().size() > 0) {
			if (this.status == BillStatus.PENDING || this.status == BillStatus.POSTED) {
				if (getTotalPayments().compareTo(getTotal()) >= 0) {
					this.setStatus(BillStatus.PAID);
					return true;
				} else if (this.status == BillStatus.PENDING) {
					this.status = BillStatus.POSTED;
				}
			}
		}
		return false;
	}

	public void removePayment(Payment payment) {
		if (payment != null && this.payments != null) {
			this.payments.remove(payment);
		}
	}

	public Set<Bill> getAdjustedBy() {
		return adjustedBy;
	}

	public void setAdjustedBy(Set<Bill> adjustedBy) {
		this.adjustedBy = adjustedBy;
	}

	public void addAdjustedBy(Bill adjustedBill) {
		checkAuthorizedToAdjust();
		if (adjustedBill == null) {
			throw new NullPointerException("The adjusted bill to add must be defined.");
		}

		if (this.adjustedBy == null) {
			this.adjustedBy = new HashSet<Bill>();
		}

		adjustedBill.setBillAdjusted(this);
		this.adjustedBy.add(adjustedBill);
	}

	public void removeAdjustedBy(Bill adjustedBill) {
		if (adjustedBill != null && this.adjustedBy != null) {
			this.adjustedBy.remove(adjustedBill);
		}
	}

	private void checkAuthorizedToAdjust() {
		if (!Context.hasPrivilege(PrivilegeConstants.ADJUST_BILLS)) {
			throw new AccessControlException("Access denied to adjust bill.");
		}
	}

	public void recalculateLineItemOrder() {
		int orderCounter = 0;
		for (BillLineItem lineItem : this.getLineItems()) {
			lineItem.setLineItemOrder(orderCounter++);
		}
	}

	public String getLastUpdated() {
		SimpleDateFormat ft = Context.getDateTimeFormat();
		String changedStr = (this.getDateChanged() != null) ? ft.format(this.getDateChanged()) : null;
		String createdStr = (this.getDateCreated() != null) ? ft.format(this.getDateCreated()) : "";
		String dateString = (changedStr != null) ? changedStr : createdStr;

		return dateString;
	}
}

File: omod_1.x/src/main/webapp/resources/js/model/bill.js
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 2.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
 * the License for the specific language governing rights and
 * limitations under the License.
 *
 * Copyright (C) OpenHMIS.  All Rights Reserved.
 */
define(
	[
		openhmis.url.backboneBase + 'js/model/generic',
		openhmis.url.cashierBase + 'js/model/cashPoint',
		openhmis.url.backboneBase + 'js/model/patient',
		openhmis.url.cashierBase + 'js/model/payment',
		openhmis.url.cashierBase + 'js/model/lineItem'
	],
	function(openhmis) {
		openhmis.Bill = openhmis.GenericModel.extend({
			trackUnsaved: true,
			meta: {
				name: openhmis.getMessage('openhmis.cashier.bill'),
				namePlural: openhmis.getMessage('openhmis.cashier.billPlural'),
				openmrsType: 'data',
				restUrl: "v2/cashier/bill"
			},
			
			schema: {
				billAdjusted: { type: 'Object', objRef: true },
				cashPoint: { type: 'Object', objRef: true },
				lineItems: { type: "List", itemType: "NestedModel", model: openhmis.LineItem },
				patient: { type: 'Object', objRef: true },
				payments: { type: "List", itemType: "NestedModel", model: openhmis.Payment},
				status: { type: 'Text' },
                adjustmentReason: {type: 'Text'}
			},
						
			BillStatus: {
				PENDING:	"PENDING",
				POSTED:		"POSTED",
				PAID:		"PAID",
				ADJUSTED:	"ADJUSTED"
			},
			
			initialize: function(attrs, options) {
				openhmis.GenericModel.prototype.initialize.call(this, attrs, options);
				if (!this.get("lineItems")) {
					this.set("lineItems", new openhmis.GenericCollection([], { model: openhmis.LineItem }), { silent: true });
				}
				if (!this.get("payments")) {
					this.set("payments", new openhmis.GenericCollection([], { model: openhmis.Payment }), { silent: true });
				}
				if (!this.get("status")) {
					this.set("status", this.BillStatus.PENDING, { silent: true });
				}
			},
			
			addPayment: function(payment) {
				this.get("payments").add(payment);
			},
			
			getTotal: function() {
				var total = 0;
				var lineItems = this.get("lineItems");
				if (lineItems && lineItems.length > 0) {
					lineItems.each(function(item) {
						if (item !== null && item.isClean()) {
							if (item.get("item ") !== null && item.get("item").id !== $('#roundingItemUuid').val()) {
								total += item.getTotal();
							}
						}
					});
				}
				return total;
			},
			
			/**
			 * This method takes the adjustment process into account when
			 * calculating the bill total
			 */
			getAdjustedTotal: function() {
				var billAdjusted = this.get("billAdjusted");
				if (billAdjusted !== undefined && this.get("status") == this.BillStatus.PENDING) {
					return this.getTotal() + billAdjusted.getTotal();
				} else {
					return this.getTotal();
				}
			},
			
			getTotalPayments: function() {
				var total = 0;
				var payments = this.get("payments");
				if (payments && payments.length > 0) {
					payments.each(function(payment) {
						if (payment !== null && payment.get("voided") !== true) {
							total += payment.get("amountTendered");
						}
					});
				}
				return total;
			},
			
			getTotalPaymentsAmount: function() {
				var total = 0;
				var payments = this.get("payments");
				if (payments && payments.length > 0) {
					payments.each(function(payment) {
						if (payment !== null && payment.get("voided") !== true) {
							total += payment.get("amount");
						}
					});
				}
				return total;
			},

			getAmountPaid: function() {
				var total = this.getTotal();
				var totalPayments = this.getTotalPayments();
				
				return Math.min(total, totalPayments);
			},
			
			getAdjustedAmountPaid: function() {
				var amount = 0;
				var billAdjusted = this.get("billAdjusted");
				if (billAdjusted !== undefined && this.get("status") == this.BillStatus.PENDING) {
					amount = billAdjusted.getTotalPaymentsAmount();
				}
				return amount;
			},
			
			validate: function(goAhead) {
				// By default, backbone validates every time we try try to alter
				// the model.  We don't want to be bothered with this until we
				// care.
                if (goAhead !== true) {
                	return null;
                }
				var lineItems = this.get("lineItems");
				if (this.get("patient") === undefined) {
					return { patient: openhmis.getMessage('openhmis.cashier.bill.choosePatientErrorMessage') }
				}
				if (lineItems === undefined || lineItems.length === 0) {
					return { lineItems: openhmis.getMessage('openhmis.cashier.bill.chooseItemErrorMessage') }
				}
				if (lineItems !== undefined && lineItems.length > 0) {
					var errors = false;
					lineItems.each(function(item) {
						if (item.attributes.quantity === 0) {
							errors = true;
						}
					});
					if (errors === true) {
						return {lineItems: openhmis.getMessage('openhmis.cashier.bill.zeroItemErrorMessage')}
					} else {
						return null;
					}
				}
				return null;
			},
			
			toJSON: function() {
				var attrs = openhmis.GenericModel.prototype.toJSON.call(this);
				if (attrs.lineItems) {
					attrs.lineItems = attrs.lineItems.toJSON();
					for (var i in attrs.lineItems) {
						attrs.lineItems[i].lineItemOrder = parseInt(i);
					}
				}
				return attrs;
			},
			
			parse: function(resp) {
				if (resp === null) {
					return resp;
				}
				if (resp.patient) {
					resp.patient = new openhmis.Patient(resp.patient);
				}
				if (resp.adjustedBy) {
					resp.adjustedBy = new openhmis.GenericCollection(resp.adjustedBy, { model: openhmis.Bill });
				}
				
				if (resp.billAdjusted) {
					resp.billAdjusted = new openhmis.Bill(resp.billAdjusted);
				} else {
					delete resp.billAdjusted;
				}
				
				if (resp.lineItems) {
					resp.lineItems = new openhmis.GenericCollection(resp.lineItems, {
						model: openhmis.LineItem,
						parse: true
					});
				}
				if (resp.payments) {
					var urlRoot = this.url() + '/payment/';
					var paymentCollection = new openhmis.GenericCollection([], { model: openhmis.Payment });
					paymentCollection.add(resp.payments, { parse: true, urlRoot: urlRoot });
					resp.payments = paymentCollection;
				}
				if (resp.cashPoint) {
					resp.cashPoint = new openhmis.CashPoint(resp.cashPoint);
				}
				return resp;
			},
			
			toString: function() {
				var receiptNumber = this.get("receiptNumber");
				var str = null;
				if (receiptNumber == null) {
					str = this.get("id");
				} else{
					str = receiptNumber;
				}
				return str ? str : openhmis.GenericModel.prototype.toString.call(this);
			}
		});
		
		return openhmis;
	}
);
